private/pin.sh

```diff
@@ -2,6 +2,7 @@
 
 set -euo pipefail
 readonly maven_install_json_loc=$BUILD_WORKSPACE_DIRECTORY/{repository_name}_install.json
+readonly workspace_name=$(basename $BUILD_WORKSPACE_DIRECTORY)
```

Comment:
Here we depend on an implementation detail of bazel (the naming of the directory an external repository is put into). While the directory name is unlikely to change in the near future, it is still good paractise to only depend on official interfaces, not implementation details, as the latter is fragile in the long run. Would providing [`repository_ctx.name`](https://docs.bazel.build/versions/master/skylark/lib/repository_ctx.html#name) to the script be an alternative to the `basename` computation?

Reply:
Yes, the `repository_ctx.name` is not necessarily the same as the directory's basename.

Reply:
I don't think `repository_ctx.name` is what we want here. `repository_ctx.name` is returning the name we define in `maven_install`, either `maven` or `unpinned_maven`. What we are looking for is the name of workspace like `rules_jvm_external` or `io_bazel`. I have limited knowledge how to get it in `coursier.bzl`.

Reply:
Hmm.. even if we get the local workspace name when pinning artifacts, the problem is that if you're exporting a `maven_install` to be loaded, the local repository may not name your external repo as the repo used in the `maven_install_json` attribute.

e.g. if you export `maven_install(maven_install_json = "@annex//:maven_install.json")` in a macro to be loaded, downstream workspaces have to name this dependency `annex` and not something else. This is tricky and I think it may require usage of repository remapping.

This also raises an important question on how to connect repositories that use `rules_jvm_external`. The [current guidance in the README](https://github.com/bazelbuild/rules_jvm_external#exporting-and-consuming-artifacts-from-external-repositories) is to load the list of artifacts from the external repo and merge them with your local `maven_install`.

An idea: If you're using your local `maven_install.json`, use `@//:maven_install.json` or just `//:maven_install.json`. If you're using a external repository's `maven_install.json`, use a function from `rules_jvm_external` that accepts the name assigned to the external repository in the local workspace, and loads some set of variables by convention from the external dependency.

```python
load("@rules_jvm_external//:defs.bzl", "external_maven_install")

http_archive(
    name = "annex",
    urls = [...],
    repo_mapping = {"@maven": "@annex_maven"},
)
# annex_foo workspace depends on `@maven//`, 
# but `@maven` may already be used by the local workspace.

# creates `@annex_maven`
external_maven_install(name = "annex_maven", for_repository = "annex")
```

where `external_maven_install` is a macro that creates a repository rule that loads `@<repository_name>//:maven_install.json` by default and names it `<repository_name>_maven`.

Reply:
Filed https://github.com/bazelbuild/rules_jvm_external/issues/212